components/director/internal/domain/application/service.go

```diff
@@ -514,6 +538,43 @@ func (s *service) genericCreate(ctx context.Context, in model.ApplicationRegiste
 	return id, nil
 }
 
+func (s *service) filterUniqueNonExistingApplications(ctx context.Context, applicationInputs []model.ApplicationRegisterInput) ([]model.ApplicationRegisterInput, error) {
+	appTenant, err := tenant.LoadFromContext(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "while loading tenant from context")
+	}
+
+	allApps, err := s.appRepo.ListAll(ctx, appTenant)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while listing all applications for tenant %s", appTenant)
+	}
+
+	var uniqueNonExistingApps []model.ApplicationRegisterInput
+	for _, ai := range applicationInputs {
+		alreadyExits := false
+
+		for _, a := range allApps {
+			if ai.Name == a.Name && ((ai.SystemNumber == nil && a.SystemNumber == nil) || (ai.SystemNumber != nil && a.SystemNumber != nil && *(ai.SystemNumber) == *(a.SystemNumber))) {
+				alreadyExits = true
+				break
+			}
+		}
+
+		for _, a := range uniqueNonExistingApps {
```

Comment:
If I understand correctly this second loop here is because it might happen so that the external system service returns duplicate system records.

If that's the case I think the code would be easier to read if `uniqueNonExistingApps` was a map[[]model.ApplicationRegInput]boolean (essentially a Set data struct). Then after the loop:

```
if !alreadyExits {
			uniqueNonExistingApps[ai] = true
		}
```

And afterwards just convert it back to an array.
I think it's better that way and it's an implementation detail that we might get duplicate systems, this should not overcomplicate our generic code.

Reply:
cannot use `model.ApplicationRegInput` as key, but perhaps a simpler key containing only the searched properties can be used. Only `systemNumber`, `Name` and `tenant`